plutus-ledger-api/test/Spec/Versions.hs

```diff
@@ -0,0 +1,95 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Spec.Versions (tests) where
+
+import PlutusCore as PLC
+import PlutusCore.Data as PLC
+import PlutusCore.MkPlc as PLC
+import PlutusLedgerApi.Common
+import PlutusLedgerApi.Common.Versions
+import PlutusLedgerApi.V1 qualified as V1
+import PlutusLedgerApi.V1.Scripts
+import PlutusLedgerApi.V2 qualified as V2
+import PlutusLedgerApi.V3 qualified as V3
+import PlutusPrelude
+import UntypedPlutusCore as UPLC
+
+import Codec.Serialise
+import Control.Monad.Except
+import Data.ByteString.Lazy as BSL
+import Data.ByteString.Short as BSS
+import Data.Either
+import Data.Foldable (for_)
+import Data.Map qualified as Map
+import Data.Set qualified as Set
+import Test.Tasty
+import Test.Tasty.HUnit
+import Test.Tasty.QuickCheck
+
+serialiseDataExScript :: SerialisedScript
+serialiseDataExScript = toShort . toStrict $ serialise serialiseDataEx
+    where
+      serialiseDataEx :: Script
+      serialiseDataEx = Script $ UPLC.Program () (PLC.defaultVersion ()) $
+                             UPLC.Apply () (UPLC.Builtin () PLC.SerialiseData) (PLC.mkConstant () $ I 1)
+
+errorScript :: SerialisedScript
+errorScript = toShort . toStrict $ serialise errorEx
+    where
+      errorEx :: Script
+      errorEx = Script $ UPLC.Program () (PLC.defaultVersion ()) $ UPLC.Error ()
+
+tests :: TestTree
+tests = testGroup "versions"
+    [ testLangVersions
+    , testBuiltinVersions
+    ]
+
+testLangVersions :: TestTree
+testLangVersions = testGroup "langs"
+    [ testProperty "v1 not before but after" $ prop_notBeforeButAfter V1.assertScriptWellFormed alonzoPV
+    , testProperty "v2 not before but after" $ prop_notBeforeButAfter V2.assertScriptWellFormed vasilPV
+    , testProperty "v3 not before but after" $ prop_notBeforeButAfter V3.assertScriptWellFormed changPV
+    , testProperty "protocol-versions can add but not remove language versions" $
+        \pvA pvB -> pvA < pvB ==> languagesAvailableIn pvA `Set.isSubsetOf` languagesAvailableIn pvB
+    ]
+  where
+    prop_notBeforeButAfter phase1 expectedPv genPv =
+        -- run phase 1 on an example script
+        let resPhase1Ex = runExcept $ phase1 genPv errorScript
+        in case compare genPv expectedPv of
```

Comment:
Could we say `if genPV < expectedPV then ...` here? (I don't really mind, I just feel obliged to make some comments to prove that I've looked at the code.)

Reply:
Indeed, thanks!